zk/legacy_executor_verifier/legacy_executor_verifier.go

```diff
@@ -264,26 +178,95 @@ func (v *LegacyExecutorVerifier) handleRequest(ctx context.Context, request *Ver
 
 	previousBlock, _ := rawdb.ReadBlockByNumber(tx, blocks[0]-1)
 
-	ok, err := execer.Verify(payload, request, previousBlock.Root())
+	// eager promise will do the work as soon as called in a goroutine, then we can retrieve the result later
+	promise := NewPromise[*VerifierResponse](func() (*VerifierResponse, error) {
+		p := payload
+		r := request
+		root := previousBlock.Root()
+		failureResponse := &VerifierResponse{
+			BatchNumber: request.BatchNumber,
+			Valid:       false,
+			Witness:     witness,
+		}
+
+		ok, err2 := execer.Verify(p, r, root)
+		if err2 != nil {
+			return failureResponse, err2
+		}
+
+		response := &VerifierResponse{
+			BatchNumber: request.BatchNumber,
+			Valid:       ok,
+			Witness:     witness,
+		}
+
+		return response, nil
+	})
+
+	// add batch to the list of batches we've added
+	v.batchesMutex.Lock()
+	v.addedBatches[request.BatchNumber] = struct{}{}
+	v.batchesMutex.Unlock()
+
+	// add the promise to the list of promises
+	v.promises = append(v.promises, promise)
+	return promise, nil
+}
+
+func writeBatchToStream(result *VerifierResponse, hdb *hermez_db.HermezDbReader, roTx kv.Tx, v *LegacyExecutorVerifier) error {
+	blks, err := hdb.GetL2BlockNosByBatch(result.BatchNumber)
 	if err != nil {
-		return false, err
+		return err
 	}
 
-	if ok {
-		// update the datastream now that we know the batch is OK
-		if err = server.WriteBlocksToStream(tx, hermezDb, v.streamServer, v.stream, blocks[0], blocks[len(blocks)-1], "verifier"); err != nil {
-			return true, err
+	if err := server.WriteBlocksToStream(roTx, hdb, v.streamServer, v.stream, blks[0], blks[len(blks)-1], "verifier"); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (v *LegacyExecutorVerifier) ConsumeResults(tx kv.RwTx) ([]*VerifierResponse, error) {
+	v.promisesLock.Lock()
+	defer v.promisesLock.Unlock()
+
+	hdb := hermez_db.NewHermezDbReader(tx)
+
+	results := make([]*VerifierResponse, len(v.promises))
+	for i, promise := range v.promises {
+		result, err := promise.Get(func(r *VerifierResponse) error {
+			return writeBatchToStream(v.promises[i].result, hdb, tx, v)
+		})
+		if err != nil {
+			log.Error("error getting verifier result", "err", err)
 		}
+		results[i] = result
 	}
 
-	response := &VerifierResponse{
-		BatchNumber: request.BatchNumber,
-		Valid:       ok,
-		Witness:     witness,
+	// clear the promises
+	v.promises = nil
+
+	return results, nil
+}
+
+func (v *LegacyExecutorVerifier) getNextOnlineAvailableExecutor() ILegacyExecutor {
+	var exec ILegacyExecutor
+
+	// TODO: find executors with spare capacity
+
+	// attempt to find an executor that is online amongst them all
+	for i := 0; i < len(v.executors); i++ {
+		v.executorNumber++
+		if v.executorNumber >= len(v.executors) {
+			v.executorNumber = 0
+		}
+		temp := v.executors[v.executorNumber]
+		if temp.CheckOnline() {
```

Comment:
If all of our executors are offline then we will have an infine loop. Do we need additional handing or keep looping over them is more than enough?

Reply:
Yeah good point - I think we can note this for the 'executor pool' followup task!